### PR TITLE
Moderated rooms or subdomains

### PR DESCRIPTION
**Pull request by damencho**
When a user joins such room or subdomain in order to be a moderator needs to provide a valid jwt token for that room.

### CODE REVIEW

File: resources/prosody-plugins/mod_muc_allowners.lua [hunk 1]
```diff
@@ -1,12 +1,99 @@
...
+        local actor_node = jid.node(actor);
+        if actor_node == 'focus' and item.attr.affiliation == 'owner' then
+            local moderated = isModerated(room_jid);
+
+            if moderated then
+                module:log('debug', 'skip focus setting owner for: %s in %s', item.attr.jid, room_jid);
+                return true;
+            end
+        end
+    end, -1); -- the priority in prosody for these is -2, we want to act before it
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_muc_allowners.lua:65]: What are "these"?

File: resources/prosody-plugins/mod_muc_allowners.lua [hunk 2]
```diff
@@ -1,12 +1,99 @@
...
+            return true, target_subdomain, nil;
+        end
+    elseif moderated_rooms:contains(room_node) then
+        return true, nil, room_node;
+    end
+
+    return false;
+end
+
+-- we need to hook on stanzas to be able to prevent focus(jicofo) from setting owner for the
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_muc_allowners.lua:36]: Is there still a valid use-case for jicofo doing that? Even if there is, can we make that configurable instead?

File: resources/prosody-plugins/mod_muc_allowners.lua [hunk 3]
```diff
@@ -1,12 +1,99 @@
...
+end
+load_config();
+
+-- Checks whether the jid is moderated, the room name is in moderated_rooms
+-- or if the subdomain is in the moderated_subdomains
+-- @return returns on of the:
+--      -> false
+--      -> true, subdomain, nil (in case of matching by subdomain)
+--      -> true, nil, room_name (if matched by room name)
+local function isModerated(room_jid)
```
Comment (bgrozev) [inline resources/prosody-plugins/mod_muc_allowners.lua:19]: is_moderated?

File: resources/prosody-plugins/mod_muc_allowners.lua [hunk 4]
```diff
@@ -1,12 +1,99 @@
...
+    -- Host room
+    "iq-set/host/http://jabber.org/protocol/muc#admin:query"
+})
+do
+    module:hook(event_name, function (event)
+        local origin, stanza = event.origin, event.stanza;
+        local room_jid = jid.bare(stanza.attr.to);
+        local actor = stanza.attr.from;
+
+        local item = stanza.tags[1].tags[1];
```
[the remark this answers is not in the document]
Comment (damencho) [reply resources/prosody-plugins/mod_muc_allowners.lua:50]: Ah ... well it comes from here :) https://github.com/bjc/prosody/blob/master/plugins/muc/muc.lib.lua#L973
This is enabled only on the main muc component and we do not enable it on breweries.

File: resources/prosody-plugins/mod_muc_allowners.lua (hunk 1 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_allowners.lua:65]: https://github.com/bjc/prosody/blob/master/plugins/muc/mod_muc.lua#L487

File: resources/prosody-plugins/mod_muc_allowners.lua [hunk 5]
```diff
@@ -1,12 +1,99 @@
...
+            module:log('debug', 'skip allowners for non-auth user subdomain:%s room_name:%s', subdomain, room_name);
+            return;
+        end
+
+        if subdomain and not subdomain == session.jitsi_meet_context_group then
+            module:log('debug', 'skip allowners for auth user and non matching room subdomain: %s, jwt subdomain: %s', subdomain, session.jitsi_meet_context_group);
+            return;
+        end
+
+        if room_name and not room_name == session.jitsi_meet_room then
```
[the remark this answers is not in the document]
Comment (damencho) [reply resources/prosody-plugins/mod_muc_allowners.lua:91]: Nope, as jwt is supposed to have room: "room_name" and you will be denied access earlier than hitting it here. The only way to come here is either have a valid room name in token or room: "*" which matches the logic.

File: resources/prosody-plugins/mod_muc_allowners.lua [hunk 6]
```diff
@@ -1,12 +1,99 @@
+local jid = require "util.jid";
```
Comment (JonathanLennox) [inline resources/prosody-plugins/mod_muc_allowners.lua:1]: Does it still make sense to call this module "allowners"?

Comment (damencho) [reply resources/prosody-plugins/mod_muc_allowners.lua:1]: Nope, we can copy it under a different name, but we should keep the old one as we will break stuff by renaming it.

**Pull request merged**